### ISSUE TITLE
quoted absolute paths being treated as relative paths when dotnet commands are run from unix bash script

### ISSUE THREAD

**Issue by Anipik** [dotnet/cli#10802]
## BackGround
I am trying to build and runtests for dotnet\machinelearning repo using  .net core 3.0 sdk preview 2 on a ubuntu machine.

This  https://github.com/dotnet/machinelearning/blob/master/init-tools.sh script is used to initialize the tools for the machine learning repo. 
This script ends up calling https://github.com/dotnet/machinelearning/blob/master/init-tools.sh#L164
with parameters with absolute paths to csproj , tools and packages directory.

This script ends up calling https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/init-tools.sh#L82

which packages path as "\machinelearning\packages\" (an absolute path) which is set here https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/init-tools.sh#L38

The expected behoviour is packages being restored in \machinelearning\packages directory but they get restored in \machinelearning\machinelearning\packages path.

## Repro with Independent Project

This behavior can be easily repro with a standalone project.

-  Create a new csproj project
-  Create a new script (restore.sh) with content as follows
```
#!/usr/bin/env bash
__PROJECT_DIR=${1:-}
__DOTNET_CMD=${2:-}
__PACKAGES_DIR=${3:-}

__DEFAULT_RESTORE_ARGS="--no-cache --packages \"${__PACKAGES_DIR}\""
__INIT_TOOLS_RESTORE_ARGS="${__DEFAULT_RESTORE_ARGS} --source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json --source https://api.nuget.org/v3/index.json ${__INIT_TOOLS_RESTORE_ARGS:-}"
$__DOTNET_CMD restore  "$__PROJECT_DIR"  $__INIT_TOOLS_RESTORE_ARGS /pp``
```
- run the restore script with ``` ./restore.sh /git/RuntimeInformation/ConsoleApp18/ConsoleApp18.csproj  /machinelearning/Tools/dotnetcli/dotnet /git/packages```

This behavior can also be repro with the dotnet build command with specifying ```--output parameter```

Note:- Everything works fine if we run this command(restore with --packages argument) from the command line. Another thing to notice if we remove the double quotes from here ```__DEFAULT_RESTORE_ARGS="--no-cache --packages \"${__PACKAGES_DIR}\""``` Then also it works fine.

This problem is not there if we change the sdk version to 2.1

On debugging a little deeper the property that is getting set wrong is ```RestorePackagesPath```.

cc @eerhardt @safern @danmosemsft @livarcocc

**Comment by peterhuene** [comment]
Hi @Anipik,

There was a breaking change to 3.0 to properly expand relative paths to a number of CLI options (see #9892).  Thus it is much more consistent with how we handle path options.

Just looking at the repro steps here, it seems you're passing `--packages /git/packages` at the option.  Is your intention that `/git/packages` should be relative to the CWD?  We might have joined this previously to the project directory in an incorrect fashion before this change, but now we combine the path given with the CWD as one would expect from a CLI tool.

Thus passing `./git/packages` or simply `git/packages` should work.  Could you give that a try?

**Comment by Anipik** [comment]
```/git/packages``` is an absolute path on unix systems. so i dont want to prepend it with CWD or project Directory.

The argument that gets passed is ``` --packages "/git/packages"```. When the argument is just "--packages /git/packages" then it works fine

**Comment by peterhuene** [comment]
Gotcha, it was intended to be a full path.  So it works without quotes then?  Could be a bug in the command line parser or how we're doing the path joining.

I'll investigate, thanks for the detailed information!

**Comment by Anipik** [comment]
> We might have joined this previously to the project directory in an incorrect fashion before this change, but now we combine the path given with the CWD as one would expect from a CLI tool

Couple of questions here, the behovior doesnt repro if the quotes are not present.  Neither does it repro when absolute path is given input from the commandline ?

> `./git/packages` or simply `git/packages` should work. Could you give that a try?

Both of these are relative paths and it works for both of them

**Comment by Anipik** [comment]
>  Gotcha, it was intended to be a full path. So it works without quotes then?

Yes it works without quotes. do you mean unix command line parser or dotnet command line parser ?
The scripts work fine with netcoreapp2.1

**Comment by peterhuene** [comment]
I meant the dotnet CLI parser.

**Comment by Anipik** [comment]
Yes it could be dotnet CLI parser.
```GetFullPath``` api treats ```\git\packages``` as an absolute path on unix and wont preprend the cwd in the return string

**Comment by peterhuene** [comment]
If I had to take an educated guess on why this is now broken it is because we're no longer passing the value from the command line (when expanded in this script, we'll get `"/git/packages"`) directly through to the MSBuild command line as `/p:RestorePackagesPath="/git/packages"`; thus, MSBuild will strip the quotes out when setting the property.

Instead, we're now trying to treat `"/git/packages"` as the path to expand before passing the full path to MSBuild.  I would imagine since quotes are valid directory characters, this would expand to `<cwd>/"/git/packages"`.

Given that this is no different from doing `dotnet restore --packages \"/git/packages\"` from the command line, I might be inclined to say won't fix if this is the case.  Still, let me confirm this hypothesis before making any decision.

**Comment by Anipik** [comment]
So if thats the case, the solution would be to remove the quotes when providing the path from the script ? in this case here https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/PackageFiles/init-tools.sh#L37

**Comment by eerhardt** [comment]
The issue with removing the quotes is when you have a ` ` space in your path... We would no longer be able to build on a machine that had a space in the path.

**Comment by peterhuene** [comment]
If you want to have the nested variables like this to build up a command line to execute and expect it to resolve arguments like you would execute it from within your terminal, I would use `eval` on the variable:

```
eval "$__DOTNET_CMD restore  "$__PROJECT_DIR"  $__INIT_TOOLS_RESTORE_ARGS /pp"
```

Thus bash will respect the double quotes around the argument to `--packages` and pass it, even if it has spaces, as a single arguments, but without the literal quotes characters that `dotnet` is trying to respect as legal path characters.

**Comment by peterhuene** [comment]
I've confirmed my hypothesis as stated above.  This is a result of the CLI now trying to expand the path argument based on CWD and your script is passing a value that has double quotes in it to the CLI.  This worked in previous versions simply because we were passing the value as-is to MSBuild, which gleefully strips the quotes from the property value despite the fact the path could start with double quotes, technically speaking.

I would suggest using the `eval` built-in supported by many different shells to cause the shell interpreter to evaluate a string as a command line and thus handling the argument quoting, rather than passing double quote literals to `dotnet` and expecting it not to interpret the literal characters as part of a path like it now does in 3.0.

I'm going to resolve this won't fix as I think the 3.0 behavior is correct and the previous behavior of it "working" was an implementation detail of how we passed the value straight through to MSBuild previously.
